### DISCUSSION TITLE
tutorials/lift-off-part2/02-exploring-our-data

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#42] [course feedback]
# tutorials/lift-off-part2/02-exploring-our-data

https://www.apollographql.com/tutorials/lift-off-part2/02-exploring-our-data

**Comment by CMH7** [comment]
How I wonder how is the API was built like if I got Apollo server with graphql then i still need to create a new server for my backend is that right? Help me got a little confused here :<

**Reply by CMH7** [reply]
Note: what I mean was if my graphql is to retrieve to another source like an API then I need to create a backend server for it, right? Isn't there any way to combine such thing a backend apollo server like I do not need to get or fetch data to another source instead directly from the apollo server or graphql server or am I just dreaming? help me sorry for confusion, still new and learning with graphql and apollo.

**Reply by whimzyLive** [reply]
Apollo server is just another server operating on top of the node.js, so yes you can directly query the records from database from the apollo server only. This example is bit misleading.

**Reply by CMH7** [reply]
So in comparison, Apollo server is just like Express server? I am using Express.js for my projects.

**Reply by whimzyLive** [reply]
@CMH7  pretty much, you can even integrate apollo with existing express server by using middleware - https://www.apollographql.com/docs/apollo-server/api/express-middleware/#expressmiddleware

**Reply by CMH7** [reply]
Ow!, hays :< I do really need to learn more about using middlewares I don't know much of them but thanks @whimzyLive for mentioning it. I'll try it.

**Reply by Jackie-Tan** [reply]
@CMH7 if I may chip in my idea, this example exhibits a use case where you are querying third-party APIs. Even better, you may be able to query third-party APIs and augment the queried data with the data from your database for example. 

By the way, I'm new to GraphQL and Apollo too! Great opportunities to learn from each other here. :)

**Reply by CMH7** [reply]
@Jackie-Tan yes, I think that maybe the use case of it. And I'm still stuck on how to create my graphql apollo server get data from database ugh. If anyone can help me, please teach me :<. My database is Firebase Firestore (not technically a fan but I just want to learn it because why not? Hehe)

**Reply by mabuyo** [reply]
Hey @CMH7! I'm not familiar with Firebase Firestore but we have an [example of a datasource using sqlite3 here](https://github.com/apollographql/odyssey-voyage-II-server/blob/main/monolith/datasources/bookings.js).

I think you could follow the same pattern. Essentially, you'll want to [initialize the connection to your database](https://firebase.google.com/docs/firestore/query-data/get-data#initialize) in the datasource constructor. 

Then, for each method, make the appropriate call to your database to get data back. For example, you might [get a specific track](https://firebase.google.com/docs/firestore/query-data/get-data#get_a_document).

Since this is outside the scope of this particular tutorial, I'll invite you to join our Discord - https://discord.gg/graphos - it could be a good topic for a stream to work together on!

**Comment by attilacsanyi** [comment]
Just a minor, there is already 7 endpoints available.
`PATCH ​/track​/{id}​/numberOfViews Increases the number of views for a track by 1` is the missing one from the screenshot.

**Comment by Hydrock** [comment]
Now on https://odyssey-lift-off-rest-api.herokuapp.com/ available 7 endpoints, but in article mentioned 6^

**Reply by ikoral** [reply]
Patch endpoint is separated into two endpoints. Rest of the endpoints are same. 
You can update both numberOfViews and numberOfLikes now. It was only numberOfViews.

**Comment by EmersonCDias** [comment]
I have a question about this

> It's okay that the array includes fields that we don't need. Our resolver functions will take care of filtering the data properties to match only what the query asks for.

Let's say our GraphQL Server is getting data from a REST API service. This REST API service has its own logic for how to retrieve data from the database.

Even though the resolver can filter just what we need, the REST API will still get all the "unnecessary" data from the database.

If it makes sense, how to solve that?

**Reply by lizhennessy** [reply]
Hey there @EmersonCDias! What you've outlined is exactly right - REST endpoints are often plagued by this kind of overfetching, where databases are taxed a lot more than they should be. This is one of the benefits of using GraphQL, not just in the API gateway layer, but as its own service, because it helps us ask for data with greater precision: no more, and no less, than what we need! 

In this course, we're querying from an existing REST endpoint because it's a pretty common methodology in web applications, and very often GraphQL is placed as a layer _in front_ of REST services. This has some benefits, as we can give our client applications a single, seamless interface they can query through, while the GraphQL layer coordinates asking each respective REST service for the pieces it needs. This doesn't overcome the overfetching problems of REST - it simply lightens the load on the client's side to know _who_ to ask for _each_ piece of data. It's certainly not the only way to use GraphQL! Resolver functions can request data from any data source - we could even rework them to circumvent the REST service entirely and go directly to the database, though that exceeds the scope of this course. :)

Here's an article on GraphQL.com that talks about [the differences between REST and GraphQL](https://graphql.com/learn/graphql-for-rest-devs/), how they're commonly plugged together to solve problems of querying for data, and the ways that we can even evolve _past_ REST APIs, and lean on GraphQL to keep our services - and application architecture! - super lean and performant from start to finish. I hope this helps, but please let us know if you have any other questions!

**Comment by Theedon** [comment]
Just a note, as of now the author ids are no more in the form cat_{number} but in the form c_{number}. That is for the exercise on this lesson, you wouldn't query for cat_9 but c_9 instead

**Reply by Theedon** [reply]
Correction: I meant for `tracks` and not `author` instead

**Reply by mabuyo** [reply]
Hey there @Theedon! To clarify for everyone, track IDs are in the format `c_9` and author IDs are in the format `cat-9`.

**Comment by bunkscene** [comment]
This is a bit off topic, but is there any info on how you built this simple api server with swagger on Heroku?  I am very interested in spinning up some small apis myself and looking for a starting point!

**Reply by mabuyo** [reply]
Hey there @bunkscene - we used Express, [swagger-ui-express](https://www.npmjs.com/package/swagger-ui-express) and a `swagger.json` file for the docs. For Heroku, we used its Github automatic deploy feature. Hope that helps!

**Comment by NishantCoder108** [comment]
### "Exploring our data" , that will be used in next lesson

- see api end points .
- explore data points
- types, query 
- Resolvers :- a function that populates data for a particular field in graphql schema
